osu.Game/Screens/Edit/EditorTimestampParser.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text.RegularExpressions;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Objects.Types;
+
+namespace osu.Game.Screens.Edit
+{
+    public static class EditorTimestampParser
+    {
+        private static readonly Regex timestamp_regex = new Regex(@"^(\d+:\d+:\d+)(?: \((\d+(?:[|,]\d+)*)\))?$", RegexOptions.Compiled);
+
+        public static string[] GetRegexGroups(string timestamp)
+        {
+            Match match = timestamp_regex.Match(timestamp);
+            return match.Success
+                ? match.Groups.Values.Where(x => x is not Match).Select(x => x.Value).ToArray()
+                : Array.Empty<string>();
+        }
+
+        public static double GetTotalMilliseconds(string timeGroup)
+        {
+            int[] times = timeGroup.Split(':').Select(int.Parse).ToArray();
+
+            Debug.Assert(times.Length == 3);
+
+            return (times[0] * 60 + times[1]) * 1_000 + times[2];
+        }
+
+        public static List<HitObject> GetSelectedHitObjects(IReadOnlyList<HitObject> editorHitObjects, string objectsGroup, double position)
```

Comment:
The inverse of this operation, i.e. generating the selection text, is part of ruleset API via:

https://github.com/ppy/osu/blob/bc9cdb4ce07fcc7fc16117121f6bb620a8094c36/osu.Game/Rulesets/Edit/HitObjectComposer.cs#L529

Examples:

https://github.com/ppy/osu/blob/bc9cdb4ce07fcc7fc16117121f6bb620a8094c36/osu.Game.Rulesets.Osu/Edit/OsuHitObjectComposer.cs#L106-L107
https://github.com/ppy/osu/blob/bc9cdb4ce07fcc7fc16117121f6bb620a8094c36/osu.Game.Rulesets.Mania/Edit/ManiaHitObjectComposer.cs#L51-L52

Due to this, I would much rather have the individual rulesets / their composers handle this parsing, too, otherwise ruleset authors have the capability of formatting selections in their own ways, but have to rely on the main game assembly actually being able to handle them.

Reply:
Right, I'm a bit lost on how could I access the current `HitObjectComposer` in either `Editor.cs` or `EditorTimestampParser.cs`

Reply:
I think that works fine, sure.